cpp/mem.c

```diff
@@ -26,10 +26,12 @@ void __far *xalloc(unsigned long size)
 	if (size <= MAX_NEAR_ALLOC)
 	{
 		p = malloc((unsigned int)size);
-		if (p == NULL)
-			return (void __far *)p;
+		fp = (void __far *)p;
```

Comment:
This cast conversion can be *very* tricky when using near and far pointer conversions, and I don't recommend it. I have seen IA16 and OWC compilers produce the wrong desired output, and the only way to know is by running `objdump86` or `wdis` on the object file to see what code was emitted.

What we're talking about here is that malloc, internally, maintains a near heap, even in large model. So in certain cases, the near NULL pointer "0", gets converted to a far pointer DS:0, which is not 0. Its a far pointer using the near data seg's segment value (DS) and the near NULL pointer, 0, when when combined is NOT 0!

My recommendation to get around this and not by bitten by the C compiler is to not try to convert NULL pointers from near to far. (I can't recall whether this is properly handled by the C library malloc when compiled for large model, which is the point).  In general, when malloc returns NULL in large model, 'p' above will be 0. So, at that point, just return 0, rather than attempt a far conversion. By all rights, the compiler should not generate any code when converting a 'char *' to a 'char far *' in large model, unless for some reason it thinks that malloc returns a near pointer (it does internally). But I'm a bit worried because I had to rewrite malloc for OWC to use `__wcnear` pointers internally - near pointers. (OWC supports the near keyword which allows for near pointers to be manipulated in large model).

Here's the code for ELKS malloc: https://github.com/ghaerr/elks/blob/master/libc/malloc/malloc.c, check out the use of `__wcnear` which is defined to `near` for OWC. Before my conversion to using near pointers, malloc failed miserably in large model. The reason is that the default data segment is managed as a near heap, and has to be, since the normal application only has a single data segment (containing .data, .bss, the heap and stack).

In order to check the the code, I use:
```
$ wdis mem.obj    # for Open Watcom
$ ia16-elf-objdump -D -r -Mi8086 mem.o   # for IA16
```
Sorry for the long comment. I will check further myself what malloc does internally, I am now a bit worried as to what exactly malloc returns (0:0 vs DS:0) in the case of Out of Memory!

Reply:
Thanks for this. I'll read carefully. I'll try to fix the near to far conversion issues.